core/block/editor/state/position.go

```diff
@@ -180,23 +173,81 @@ func (s *State) moveFromSide(target, parent simple.Block, pos model.BlockPositio
 	if pos == model.Block_Right {
 		columnPos += 1
 	}
-	row.Model().ChildrenIds = slice.Insert(row.Model().ChildrenIds, columnPos, column.Model().Id)
+	row.Model().ChildrenIds = s.cacheInsertMoveFromSide(row.Model(), columnPos, column.Model())
 	s.changesStructureIgnoreIds = append(s.changesStructureIgnoreIds, "cd-"+opId, "ct-"+opId, "r-"+opId, row.Model().Id)
 	return
 }
 
 func (s *State) wrapToRow(opId string, parent, b simple.Block) (row simple.Block, err error) {
-	column := simple.New(&model.Block{
-		Id:          "ct-" + opId,
-		ChildrenIds: []string{b.Model().Id},
+	column := s.cacheWrapToColumn(opId, b)
+	s.Add(column)
+	row = s.cacheWriteToRaw(opId, column)
+	s.Add(row)
+	pos := slice.FindPos(parent.Model().ChildrenIds, b.Model().Id)
+	if pos == -1 {
+		return nil, fmt.Errorf("creating row: can't find child[%s] in given parent[%s]", b.Model().Id, parent.Model().Id)
+	}
+	parent.Model().ChildrenIds[pos] = row.Model().Id
+	return
+}
+
+func (s *State) cacheParent(parent *model.Block, childrenIds []string) {
+	s.cacheParentUntilSame(parent, childrenIds, false)
+}
+
+func (s *State) cacheParentUntilSame(parent *model.Block, childrenIds []string, untilSame bool) {
+	parent.ChildrenIds = childrenIds
+	if s.isIdsCacheInited() {
+		cache := s.getSubIdsCache()
+		for _, childId := range childrenIds {
+			if untilSame {
+				if curParentId, ok := cache[childId]; ok && curParentId == parent.Id {
+					break
+				}
+			}
+			cache[childId] = parent.Id
+		}
+	}
+}
+
+func (s *State) cacheAppendStart(target simple.Block, ids []string) []string {
+	result := append(target.Model().ChildrenIds, ids...)
+	s.cacheParent(target.Model(), result)
+	return result
+}
+
+func (s *State) cacheAppendEnd(target *model.Block, ids []string) {
+	result := append(ids, target.ChildrenIds...)
+	s.cacheParent(target, result)
+}
+
+func (s *State) cacheInsert(target *model.Block, pos int, ids []string) {
+	result := slice.Insert(target.ChildrenIds, pos, ids...)
+	s.cacheParent(target, result)
+}
+
+func (s *State) cacheColumnCreation(opId string, ids []string) simple.Block {
+	result := simple.New(&model.Block{
+		Id:          "cd-" + opId,
+		ChildrenIds: ids,
 		Content: &model.BlockContentOfLayout{
 			Layout: &model.BlockContentLayout{
 				Style: model.BlockContentLayout_Column,
 			},
 		},
 	})
-	s.Add(column)
-	row = simple.New(&model.Block{
+	s.cacheParent(result.Model(), ids)
```

Comment:
Isn't a new block should be cached in `st.Add`?

Reply:
yep